File: internal/vault/static/card.js
// Card network detection and validation (§2.3).
//
// A direct mirror of internal/payments/card.go. The two must agree: the browser
// uses this for instant feedback while typing, and the vault re-validates
// server-side because client-side checks are a UX feature, never a control.
//
// Entirely offline — no API key, no network call. This is the same logic
// Stripe Elements ships to browsers.

export const NETWORKS = {
  VISA: 'visa',
  MASTERCARD: 'mastercard',
  AMEX: 'amex',
  DISCOVER: 'discover',
  RUPAY: 'rupay',
  UNKNOWN: 'unknown',
};

// Ordering is load-bearing: RuPay overlaps Discover on 65 and 652x, so the
// more specific RuPay ranges are tested first. Keep in sync with card.go.
const RULES = [
  { network: NETWORKS.AMEX, digits: 2, ranges: [[34, 34], [37, 37]], lengths: [15] },
  { network: NETWORKS.RUPAY, digits: 3, ranges: [[508, 508], [606, 606], [652, 653], [817, 819]], lengths: [16] },
  { network: NETWORKS.VISA, digits: 1, ranges: [[4, 4]], lengths: [13, 16, 19] },
  { network: NETWORKS.MASTERCARD, digits: 2, ranges: [[51, 55]], lengths: [16] },
  { network: NETWORKS.MASTERCARD, digits: 4, ranges: [[2221, 2720]], lengths: [16] },
  { network: NETWORKS.DISCOVER, digits: 4, ranges: [[6011, 6011], [6221, 6229], [6440, 6499], [6500, 6599]], lengths: [16, 19] },
];

// Formatting groups per network. Amex is 4-6-5; everything else is 4-4-4-4.
const GROUPS = {
  [NETWORKS.AMEX]: [4, 6, 5],
  default: [4, 4, 4, 4],
};

export function onlyDigits(value) {
  return (value || '').replace(/\D/g, '');
}

// Works on partial input, which is what lets the network logo appear before the
// user has finished typing.
export function detectNetwork(value) {
  const digits = onlyDigits(value);
  if (!digits) return NETWORKS.UNKNOWN;

  for (const rule of RULES) {
    if (digits.length < rule.digits) continue;
    const prefix = parseInt(digits.slice(0, rule.digits), 10);
    for (const [lo, hi] of rule.ranges) {
      if (prefix >= lo && prefix <= hi) return rule.network;
    }
  }
  return NETWORKS.UNKNOWN;
}

// Mod-10 checksum. Catches typos and transpositions; says nothing about whether
// the card exists or has funds.
export function luhn(value) {
  const digits = onlyDigits(value);
  if (digits.length < 12) return false;

  let sum = 0;
  let double = false;
  for (let i = digits.length - 1; i >= 0; i--) {
    let d = digits.charCodeAt(i) - 48;
    if (double) {
      d *= 2;
      if (d > 9) d -= 9;
    }
    sum += d;
    double = !double;
  }
  return sum % 10 === 0;
}

export function lengthValid(network, length) {
  for (const rule of RULES) {
    if (rule.network === network && rule.lengths.includes(length)) return true;
  }
  return false;
}

// maxLength lets the input stop accepting digits at the right point per network.
export function maxLength(network) {
  const lengths = RULES.filter((r) => r.network === network).flatMap((r) => r.lengths);
  return lengths.length ? Math.max(...lengths) : 19;
}

export function formatNumber(value, network) {
  const digits = onlyDigits(value).slice(0, maxLength(network));
  const groups = GROUPS[network] || GROUPS.default;

  const parts = [];
  let offset = 0;
  for (const size of groups) {
    if (offset >= digits.length) break;
    parts.push(digits.slice(offset, offset + size));
    offset += size;
  }
  // Anything past the defined groups (19-digit Visa) trails on the end.
  if (offset < digits.length) parts.push(digits.slice(offset));

  return parts.join(' ');
}

// A card is valid through the last day of its stated expiry month.
export function expiryValid(month, year) {
  if (!(month >= 1 && month <= 12)) return false;

  const now = new Date();
  const currentYear = now.getFullYear();
  const fullYear = year < 100 ? 2000 + year : year;

  if (fullYear < currentYear) return false;
  if (fullYear === currentYear && month < now.getMonth() + 1) return false;
  // Reject implausibly distant dates, which are almost always typos.
  return fullYear <= currentYear + 20;
}

export function cvcLength(network) {
  return network === NETWORKS.AMEX ? 4 : 3;
}

export function validateCard({ number, expMonth, expYear, cvc }) {
  const errors = {};
  const network = detectNetwork(number);
  const digits = onlyDigits(number);

  if (!digits) {
    errors.number = 'Card number is required';
  } else if (!luhn(digits)) {
    errors.number = 'Card number is not valid';
  } else if (network !== NETWORKS.UNKNOWN && !lengthValid(network, digits.length)) {
    errors.number = 'Card number is the wrong length';
  }

  if (!expiryValid(expMonth, expYear)) {
    errors.expiry = 'Expiry date is not valid';
  }

  const wanted = cvcLength(network);
  if (onlyDigits(cvc).length !== wanted) {
    errors.cvc = `Security code must be ${wanted} digits`;
  }

  return { valid: Object.keys(errors).length === 0, errors, network };
}

File: internal/payments/card.go
package payments

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"strings"
)

// Card network detection and validation (§2.3).
//
// This is fully deterministic, offline, and genuinely production-grade — the
// same logic Stripe Elements ships to browsers. It needs no API key and no
// network call, which is why the design doc calls out that it should be built
// for real rather than stubbed. The same rules are mirrored in the JS checkout
// widget so the card logo appears while the user is still typing.

type Network string

const (
	NetworkVisa       Network = "visa"
	NetworkMastercard Network = "mastercard"
	NetworkAmex       Network = "amex"
	NetworkDiscover   Network = "discover"
	NetworkRuPay      Network = "rupay"
	NetworkUnknown    Network = "unknown"
)

var (
	ErrInvalidCardNumber = errors.New("payments: card number failed Luhn check")
	ErrInvalidLength     = errors.New("payments: card number has an invalid length for its network")
)

// rule describes one network's IIN prefixes and permitted lengths.
type rule struct {
	network Network
	// ranges are inclusive numeric prefix ranges compared at equal digit width,
	// which handles both single prefixes (4) and spans (2221-2720) uniformly.
	ranges  [][2]int
	digits  int // prefix width the ranges are expressed in
	lengths []int
}

// Order matters: RuPay overlaps Discover on 65 and 6521-6529, so the more
// specific RuPay ranges are evaluated first. Real issuers resolve this with
// full BIN tables; for detection purposes the ordering is sufficient.
var rules = []rule{
	{NetworkAmex, [][2]int{{34, 34}, {37, 37}}, 2, []int{15}},
	{NetworkRuPay, [][2]int{{508, 508}, {606, 606}, {652, 653}, {817, 819}}, 3, []int{16}},
	{NetworkVisa, [][2]int{{4, 4}}, 1, []int{13, 16, 19}},
	{NetworkMastercard, [][2]int{{51, 55}}, 2, []int{16}},
	{NetworkMastercard, [][2]int{{2221, 2720}}, 4, []int{16}},
	{NetworkDiscover, [][2]int{{6011, 6011}, {6221, 6229}, {6440, 6499}, {6500, 6599}}, 4, []int{16, 19}},
}

// DetectNetwork identifies the card network from as few digits as are needed.
// It works on partial input, which is what lets the checkout widget light up
// the network logo before the user finishes typing.
func DetectNetwork(number string) Network {
	digits := onlyDigits(number)
	if digits == "" {
		return NetworkUnknown
	}
	for _, r := range rules {
		if len(digits) < r.digits {
			continue
		}
		prefix := parseInt(digits[:r.digits])
		for _, span := range r.ranges {
			if prefix >= span[0] && prefix <= span[1] {
				return r.network
			}
		}
	}
	return NetworkUnknown
}

// Luhn validates the mod-10 checksum. It catches transposed and mistyped
// digits; it says nothing about whether the card exists or has funds.
func Luhn(number string) bool {
	digits := onlyDigits(number)
	if len(digits) < 12 {
		return false
	}
	sum, double := 0, false
	for i := len(digits) - 1; i >= 0; i-- {
		d := int(digits[i] - '0')
		if double {
			if d *= 2; d > 9 {
				d -= 9
			}
		}
		sum += d
		double = !double
	}
	return sum%10 == 0
}

// ValidateNumber checks the checksum and that the length suits the network.
func ValidateNumber(number string) (Network, error) {
	digits := onlyDigits(number)
	network := DetectNetwork(digits)

	if !Luhn(digits) {
		return network, ErrInvalidCardNumber
	}
	if network != NetworkUnknown && !lengthValid(network, len(digits)) {
		return network, ErrInvalidLength
	}
	return network, nil
}

func lengthValid(network Network, length int) bool {
	for _, r := range rules {
		if r.network != network {
			continue
		}
		for _, l := range r.lengths {
			if l == length {
				return true
			}
		}
	}
	return false
}

// Fingerprint derives a stable, non-reversible identifier for a card.
//
// HMAC rather than a bare hash: the card number space is small enough to
// enumerate (a plain SHA-256 of a 16-digit PAN is brute-forceable in minutes),
// so the secret salt is what actually prevents recovering the PAN from a
// leaked fingerprint table. This value is what velocity rules and blocklists
// key on, and it never lets us reconstruct the card (§14.5).
func Fingerprint(number, salt string) string {
	mac := hmac.New(sha256.New, []byte(salt))
	mac.Write([]byte(onlyDigits(number)))
	return "card_" + hex.EncodeToString(mac.Sum(nil))[:32]
}

// Last4 returns the final four digits, the only part of a PAN safe to store
// and display (§13).
func Last4(number string) string {
	digits := onlyDigits(number)
	if len(digits) < 4 {
		return ""
	}
	return digits[len(digits)-4:]
}

// BIN returns the first six digits, used for issuer risk lists (§14.2).
func BIN(number string) string {
	digits := onlyDigits(number)
	if len(digits) < 6 {
		return ""
	}
	return digits[:6]
}

func onlyDigits(s string) string {
	var b strings.Builder
	b.Grow(len(s))
	for _, r := range s {
		if r >= '0' && r <= '9' {
			b.WriteRune(r)
		}
	}
	return b.String()
}

func parseInt(s string) int {
	n := 0
	for _, r := range s {
		n = n*10 + int(r-'0')
	}
	return n
}
